# Mini Project2
### 20181531 김연주

In [4]:
import numpy as np
import pandas as pd

df = pd.read_csv("csv/train.csv")
df

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
0              0.288585          -0.020294          -0.132905   
1              0.278419          -0.016411          -0.123520   
2              0.279653          -0.019467          -0.113462   
3              0.279174          -0.026201          -0.123283   
4              0.276629          -0.016570          -0.115362   
...                 ...                ...                ...   
7347           0.299665          -0.057193          -0.181233   
7348           0.273853          -0.007749          -0.147468   
7349           0.273387          -0.017011          -0.045022   
7350           0.289654          -0.018843          -0.158281   
7351           0.351503          -0.012423          -0.203867   

      tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
0            -0.995279         -0.983111         -0.913526         -0.995112   
1            -0.998245         -0.975300         -0.960322         -0.998807   
2            -0.995380         -0.967187         -0.978944         -0.996520   
3            -0.996091         -0.983403         -0.990675         -0.997099   
4            -0.998139         -0.980817         -0.990482         -0.998321   
...                ...               ...               ...               ...   
7347         -0.195387          0.039905          0.077078         -0.282301   
7348         -0.235309          0.004816          0.059280         -0.322552   
7349         -0.218218         -0.103822          0.274533         -0.304515   
7350         -0.219139         -0.111412          0.268893         -0.310487   
7351         -0.269270         -0.087212          0.177404         -0.377404   

      tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  \
0            -0.983185         -0.923527         -0.934724  ...   
1            -0.974914         -0.957686         -0.943068  ...   
2            -0.963668         -0.977469         -0.938692  ...   
3            -0.982750         -0.989302         -0.938692  ...   
4            -0.979672         -0.990441         -0.942469  ...   
...                ...               ...               ...  ...   
7347          0.043616          0.060410          0.210795  ...   
7348         -0.029456          0.080585          0.117440  ...   
7349         -0.098913          0.332584          0.043999  ...   
7350         -0.068200          0.319473          0.101702  ...   
7351         -0.038678          0.229430          0.269013  ...   

      fBodyBodyGyroJerkMag-kurtosis()  angle(tBodyAccMean,gravity)  \
0                           -0.710304                    -0.112754   
1                           -0.861499                     0.053477   
2                           -0.760104                    -0.118559   
3                           -0.482845                    -0.036788   
4                           -0.699205                     0.123320   
...                               ...                          ...   
7347                        -0.880324                    -0.190437   
7348                        -0.680744                     0.064907   
7349                        -0.304029                     0.052806   
7350                        -0.344314                    -0.101360   
7351                        -0.740738                    -0.280088   

      angle(tBodyAccJerkMean),gravityMean)  angle(tBodyGyroMean,gravityMean)  \
0                                 0.030400                         -0.464761   
1                                -0.007435                         -0.732626   
2                                 0.177899                          0.100699   
3                                -0.012892                          0.640011   
4                                 0.122542                          0.693578   
...                                    ...                               ...   
7347                              0.829718                

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7352 entries, 0 to 7351
Columns: 563 entries, tBodyAcc-mean()-X to Activity
dtypes: float64(561), int64(1), object(1)
memory usage: 31.6+ MB


In [6]:
from sklearn.utils import shuffle
df = shuffle(df)

In [7]:
from sklearn.preprocessing import LabelEncoder

df_cat = df['Activity']
encoder = LabelEncoder()
encoder.fit(df_cat)
df_encoded = encoder.transform(df_cat)

In [8]:
df['Activity'] = df_encoded
df['Activity']

6389    3
1043    1
6102    1
2746    1
1237    0
       ..
2332    4
6227    5
4370    5
6977    2
7153    4
Name: Activity, Length: 7352, dtype: int32

In [12]:
X = df.drop(['subject', 'Activity'], axis=1)
y = df['Activity']

## ExtraTreesClassifier

In [13]:
from sklearn.ensemble import ExtraTreesClassifier

extra_clf = ExtraTreesClassifier(random_state=42)
extra_clf.fit(X, y)

ExtraTreesClassifier(random_state=42)

In [14]:
from sklearn.model_selection import cross_val_score

extra_score = cross_val_score(extra_clf, X, y, cv=10, scoring="accuracy") # 교차검증
extra_score

array([0.98777174, 0.97961957, 0.98911565, 0.98911565, 0.9877551 ,
       0.98095238, 0.9877551 , 0.98639456, 0.99183673, 0.98639456])

In [15]:
extra_score.mean()

0.9866711032238984

## VotingClassifier

In [20]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.ensemble import VotingClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import BaggingClassifier

log_clf = LogisticRegression(solver="liblinear", random_state=42)
rnd_clf = RandomForestClassifier(n_estimators=100, random_state=42)
svm_clf = SVC(kernel="linear", probability=True, random_state=42)
ada_clf = AdaBoostClassifier(RandomForestClassifier(random_state=42))
gbrt_clf = GradientBoostingClassifier(random_state=42)
bag_clf = BaggingClassifier(
    RandomForestClassifier(random_state=42), bootstrap=False, random_state=42)

In [21]:
def display_scores(scores):
    print("Scores:", scores)
    print("Mean:", scores.mean())
    print("Standard deviation", scores.std())

In [22]:
log_clf.fit(X, y)
rnd_clf.fit(X, y)
svm_clf.fit(X, y)
ada_clf.fit(X, y)
gbrt_clf.fit(X, y)
bag_clf.fit(X, y)

BaggingClassifier(base_estimator=RandomForestClassifier(random_state=42),
                  bootstrap=False, random_state=42)

In [23]:
log_score = cross_val_score(log_clf, X, y, cv=10, scoring="accuracy")
log_score

array([0.98777174, 0.98777174, 0.98367347, 0.99183673, 0.98503401,
       0.98503401, 0.98503401, 0.9877551 , 0.98503401, 0.98367347])

In [24]:
display_scores(log_score)

Scores: [0.98777174 0.98777174 0.98367347 0.99183673 0.98503401 0.98503401
 0.98503401 0.9877551  0.98503401 0.98367347]
Mean: 0.9862618308192843
Standard deviation 0.002393709017634704


In [26]:
rnd_score = cross_val_score(rnd_clf, X, y, cv=10, scoring="accuracy")
rnd_score

array([0.98777174, 0.97690217, 0.98231293, 0.98367347, 0.98367347,
       0.97414966, 0.98503401, 0.98367347, 0.9877551 , 0.97823129])

In [27]:
display_scores(rnd_score)

Scores: [0.98777174 0.97690217 0.98231293 0.98367347 0.98367347 0.97414966
 0.98503401 0.98367347 0.9877551  0.97823129]
Mean: 0.9823177314404022
Standard deviation 0.004300559575916856


In [29]:
svm_score = cross_val_score(svm_clf, X, y, cv=10, scoring="accuracy")
svm_score

array([0.98641304, 0.99048913, 0.9877551 , 0.98639456, 0.98231293,
       0.9877551 , 0.9877551 , 0.98911565, 0.98639456, 0.98367347])

In [30]:
display_scores(svm_score)

Scores: [0.98641304 0.99048913 0.9877551  0.98639456 0.98231293 0.9877551
 0.9877551  0.98911565 0.98639456 0.98367347]
Mean: 0.9868058636498077
Standard deviation 0.0022824494107555747


In [32]:
ada_clf_score = cross_val_score(ada_clf, X, y, cv=10, scoring="accuracy")
ada_clf_score

array([0.98641304, 0.97690217, 0.98095238, 0.98367347, 0.9877551 ,
       0.97142857, 0.97687075, 0.98367347, 0.98095238, 0.97687075])

In [33]:
display_scores(ada_clf_score)

Scores: [0.98641304 0.97690217 0.98095238 0.98367347 0.9877551  0.97142857
 0.97687075 0.98367347 0.98095238 0.97687075]
Mean: 0.9805492088139603
Standard deviation 0.004792761118387135


In [38]:
gbrt_clf_score = cross_val_score(gbrt_clf, X, y, cv=10, scoring="accuracy")
gbrt_clf_score

array([0.98641304, 0.98777174, 0.98367347, 0.99183673, 0.98639456,
       0.98231293, 0.99047619, 0.98367347, 0.99183673, 0.98095238])

In [39]:
display_scores(gbrt_clf_score)

Scores: [0.98641304 0.98777174 0.98367347 0.99183673 0.98639456 0.98231293
 0.99047619 0.98367347 0.99183673 0.98095238]
Mean: 0.9865341245193731
Standard deviation 0.003724005149009361


In [35]:
bag_clf_score = cross_val_score(bag_clf, X, y, cv=10, scoring="accuracy")
bag_clf_score

array([0.98641304, 0.97826087, 0.98231293, 0.98639456, 0.98911565,
       0.97414966, 0.98367347, 0.98639456, 0.98095238, 0.98367347])

In [36]:
display_scores(bag_clf_score)

Scores: [0.98641304 0.97826087 0.98231293 0.98639456 0.98911565 0.97414966
 0.98367347 0.98639456 0.98095238 0.98367347]
Mean: 0.9831340579710144
Standard deviation 0.004222275933630834


- LogisticRegression: 0.9862618308192843
- RandomForestClassifier: 0.9823177314404022
- SVC: 0.9868058636498077
- AdaBoostClassifier: 0.9805492088139603
- GradientBoostingClassifier:0.9865341245193731 
- BaggingClassifier: 0.9831340579710144

### SVM의 결과가 제일 좋게 나왔음

## 결과: 0.9868058636498077 => 약 98.7%